Common/Utils.hs

```diff
@@ -151,6 +154,54 @@ trimLeft = dropWhile isSpace
 trimRight :: String -> String
 trimRight = foldr (\ c cs -> if isSpace c && null cs then [] else c : cs) ""
 
+-- | convert to lower case
+downcase :: String -> String
+downcase = map toLower
+
+{-
+   Split a 'String', by specified predicate, but do not remove matched
+   characters from the result.
+
+   Copied from https://gist.github.com/ruthenium/3715696
+-}
+splitR :: (Char -> Bool) -> String -> [String]
+splitR _ [] = []
+splitR p s =
+  let
+    go :: Char -> String -> [String]
+    go m s' = case break p s' of
+      (b', []) -> [m : b']
+      (b', x : xs) -> (m : b') : go x xs
+  in case break p s of
+    (b, []) -> [b]
+    ([], h : t) -> go h t
+    (b, h : t) -> b : go h t
+
```

Comment:
Yes, `splitBy` is not usable, because it removes the delimiter. `splitR` is only used with the special predicate isUpper.

It seems easier to write a single recursive function that can replace `glueSingletons "" . splitR isUpper`.